src/content/en/updates/posts/2016/03/web-enabled-internet-of-things.markdown

```diff
@@ -0,0 +1,545 @@
+method and encode the value.
+
+{% highlight javascript %}
+TemperatureCharacteristic.prototype.onReadRequest = function(offset, callback) {
+  var data = new Buffer(8);
+  data.writeDoubleLE(this._lastValue, 0);
+  callback(this.RESULT_SUCCESS, data);
+};
+{% endhighlight %}
+
+For "notify" we need to add a method to handle subscription and
+unsubscription. Basically, we simply store a callback. When we
+have a new temperature reason we want to send, we then call that
+callback with the new value (encoded as above).
+
+{% highlight javascript %}
+TemperatureCharacteristic.prototype.onSubscribe = function(maxValueSize, updateValueCallback) {
+  console.log("Subscribed to temperature change.");
+  this._onChange = updateValueCallback;
+  this._lastValue = undefined;
+};
+
+TemperatureCharacteristic.prototype.onUnsubscribe = function() {
+  console.log("Unsubscribed to temperature change.");
+  this._onChange = null;
+};
+{% endhighlight %}
+
+As values can fluctuate a bit, we need to smoothen the values we
+get from the TMP36 sensor. I opted to simply take the average over
+100 samples and only send updates when the temperature changes by
+at least 1 degree.
+
+{% highlight javascript %}
+TemperatureCharacteristic.prototype.valueChange = function(value) {
+  this._total += value;
+  this._samples++;
+    
+  if (this._samples < NO_SAMPLES) {
+    return;
+  }
+        
+  var newValue = Math.round(this._total / NO_SAMPLES);
+    
+  this._total = 0;
+  this._samples = 0;
+    
+  if (this._lastValue && Math.abs(this._lastValue - newValue) < 1) {
+    return;
+  }
+    
+  this._lastValue = newValue;
+    
+  console.log(newValue);
+  var data = new Buffer(8);
+  data.writeDoubleLE(newValue, 0);
+    
+  if (this._onChange) {
+    this._onChange(data);
+  }
+};
+{% endhighlight %}
+
+Now, that was the temperature sensor. The color LED is even more
+simple. Below you find the object as well as the "read" method.
+The characteristic is configured to allow for "read" and "write"
+operations and has a different UUID than the temperature characteristic.
+
+{% highlight javascript %}
+var ColorCharacteristic = function() {
+  bleno.Characteristic.call(this, {
+    uuid: 'fc0b',
+    properties: ['read', 'write'],
+    value: null
+  });
+  this._value = 'ffffff';
+  this._led = null;
+};
+
+util.inherits(ColorCharacteristic, bleno.Characteristic);
+
+ColorCharacteristic.prototype.onReadRequest = function(offset, callback) {
+  var data = new Buffer(this._value);
+  callback(this.RESULT_SUCCESS, data);
+};
+{% endhighlight %}
+
+As I need to control the LED from the object, I add a
+`this._led` member which I use to store the Johnny-Five LED
+object. I also set the color of the LED to its' default
+value (white, aka `#ffffff`).
+
+{% highlight javascript %}
+board.on("ready", function() {
+  ...
+  colorCharacteristic._led = led;
+  led.color(colorCharacteristic._value);
+  led.intensity(30);
```

Comment:
It doesn't really matter and depends on your LED. Mine was annoying me as it is right on my desktop :)